File: app/src/main/java/com/example/konye/lingo/utils/schedule/SchedulerProvider.java
package com.example.konye.lingo.utils.schedule;

import io.reactivex.Scheduler;
import io.reactivex.schedulers.Schedulers;

public class SchedulerProvider implements BaseScheduler {

    private static SchedulerProvider INSTANCE = null;

    private synchronized static void createInstance() {
        if (INSTANCE == null) {
            INSTANCE = new SchedulerProvider();
        }
    }

    public static SchedulerProvider getInstance() {
        if (INSTANCE == null) createInstance();
        return INSTANCE;
    }

    @Override
    public Scheduler computation() {
        return Schedulers.computation();
    }

    @Override
    public Scheduler io() {
        return Schedulers.io();
    }

    @Override
    public Scheduler ui() {
        return AndroidSchedulers.mainThread();
    }
}

File: app/src/main/java/com/mahadum360/mahadum/utils/schedule/SchedulerProvider.kt
package com.mahadum360.mahadum.utils.schedule

import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class SchedulerProvider : BaseScheduler {

    override fun computation(): Scheduler {
        return Schedulers.computation()
    }

    override fun io(): Scheduler {
        return Schedulers.io()
    }

    override fun ui(): Scheduler {
        return AndroidSchedulers.mainThread()
    }

    companion object {

        private var INSTANCE: SchedulerProvider? = null

        @Synchronized
        private fun createInstance() {
            if (INSTANCE == null) {
                INSTANCE = SchedulerProvider()
            }
        }

        val instance: SchedulerProvider
            get() {
                if (INSTANCE == null) createInstance()
                return INSTANCE!!
            }
    }
}
